packages/loopring_v3/contracts/amm/AmplifiedAmmController.sol

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
+pragma solidity ^0.7.0;
+
+import "./IAmmController.sol";
+import "../amm/LoopringAmmPool.sol";
+import "../lib/Claimable.sol";
+import "../lib/MathUint.sol";
+import "../lib/MathUint96.sol";
+import "../thirdparty/SafeCast.sol";
+
+
+/// @author Brecht Devos - <brecht@loopring.org>
+contract AmplifiedAmmController is IAmmController, Claimable
+{
+    using MathUint          for uint;
+    using MathUint96        for uint96;
+    using SafeCast          for uint;
+
+    uint public constant AMPLIFICATION_FACTOR_BASE = (10 ** 18);
+
+    mapping(address => uint) public amplificationFactors;
+
+    function getInitialVirtualBalances(
+        uint96[] memory joinAmounts
+        )
+        external
+        view
+        override
+        returns (uint96[] memory)
+    {
+        uint amplificationFactor = getAmplificationFactor(msg.sender);
```

Comment:
I only really implemented to special case when `AF == 0` ->  `AF =1` is so that all existing pools automatically keep the existing behavior without having to explicitly set the value to `0` in this contract. But setting the virtual balance of a token to `0` will disable the AMM in the circuit for the token (similar to previously how setting the weight to `0` also disabled AMM functionality for the pool with v1). So virtual balances being 0 is allowed, the pool just cannot be used when that's the case.

I don't really see how an AMM with virtual balance 0 or even very close to zero would be useful (unless for making it inactive)? A pool with the virtual balance set to `0` (or very close to `0`) would have infinite price.

Reply:
Virtual balance being close to 0s doesn't make sense to me. I even think the virtual balances shall always be >= the actual balances, but I'm not sure if this is explicitly guaranteed/checked anywhere.

Reply:
It's not checked anywhere because it's valid, but I think virtual balances >= actual balances will indeed always be true in practice because otherwise some funds in the pool would be impossible to use.